cwrapper/CArachneWrapperTest.cc

```diff
@@ -0,0 +1,135 @@
+    MockSyscall* sys;
+
+    std::thread* coreArbiterServerThread;
+    // Override this to define how to set up the environment.
+    virtual void SetUp() {
+        // Initalize core arbiter server
+        CoreArbiter::Logger::setLogLevel(CoreArbiter::WARNING);
+        sys = new MockSyscall();
+        sys->callGeteuid = false;
+        sys->geteuidResult = 0;
+        CoreArbiterServer::testingSkipCpusetAllocation = true;
+
+        CoreArbiterServer::sys = sys;
+        coreArbiterServer = new CoreArbiterServer(std::string(TEST_SOCKET),
+                                                  std::string(TEST_MEM),
+                                                  {1, 2, 3, 4, 5, 6, 7}, false);
+        coreArbiterServerThread =
+            new std::thread([=] { coreArbiterServer->startArbitration(); });
+    }
+    // Override this to define how to tear down the environment.
+    virtual void TearDown() {
+        coreArbiterServer->endArbitration();
+        coreArbiterServerThread->join();
+        delete coreArbiterServerThread;
+        delete coreArbiterServer;
+        delete sys;
+    }
+};
+
+::testing::Environment* const testEnvironment =
+    (useCoreArbiter) ? ::testing::AddGlobalTestEnvironment(new Environment)
+                     : NULL;
+
+struct ArachneTest : public ::testing::Test {
+    virtual void SetUp() {
+        Arachne::minNumCores = 1;
+        Arachne::maxNumCores = 3;
+        Arachne::disableLoadEstimation = true;
+        ::cArachneInit(NULL, NULL);
+        // Artificially wake up all threads for testing purposes
+        std::vector<uint32_t> coreRequest({3, 0, 0, 0, 0, 0, 0, 0});
+        coreArbiter->setRequestedCores(coreRequest);
+        limitedTimeWait([]() -> bool { return numActiveCores == 3; });
+    }
+
+    virtual void TearDown() {
+        // Unblock all cores so they can shut down and be joined.
+        coreArbiter->setRequestedCores(
+            {Arachne::maxNumCores, 0, 0, 0, 0, 0, 0, 0});
+
+        ::cArachneShutDown();
+        ::cArachneWaitForTermination();
+    }
+};
+
+// Helper function for tests with timing dependencies, so that we wait for a
+// finite amount of time in the case of a bug causing an infinite loop.
+static void
+limitedTimeWait(std::function<bool()> condition, int numIterations) {
+    for (int i = 0; i < numIterations; i++) {
+        if (condition()) {
+            break;
+        }
+        usleep(1000);
+    }
+    // We use assert here because an infinite loop will result in TearDown
+    // not being able to complete, so we might as well terminate the tests here.
+    ASSERT_TRUE(condition());
+}
+
+static void*
+wrapperMain(void* arg) {
+    *(unsigned*)arg = 0xDEADBEEF;
+    return NULL;
+}
+
+TEST_F(ArachneTest, CWrapper_createThread) {
+    CArachneThreadId threadId;
+    unsigned arg = 0;
+    int ret = cArachneCreateThread(&threadId, wrapperMain, (void*)&arg);
+    EXPECT_EQ(0, ret);
+    EXPECT_EQ(0xDEADBEEF, arg);
```

Comment:
This looks like it could race with `wrapperMain`. What happens if `wrapperMain` hasn't run before this check?

I'm not sure if the second commit addresses this, but you may want to fix it in this commit since we're seeing it here.